okapi-core/src/main/java/org/folio/okapi/util/TokenCache.java

```diff
@@ -0,0 +1,104 @@
+package org.folio.okapi.util;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.OkapiLogger;
+
+public class TokenCache {
+
+  public static final long DEFAULT_TTL = 3 * 60 * 1000L;
+  
+  private final Map<String, CacheEntry> cache = new HashMap<>();
+  private final long ttl;
+  
+  private static final Logger logger = OkapiLogger.get(TokenCache.class);
+  
+  public TokenCache() {
+    this(DEFAULT_TTL);
+  }
+  
+  public TokenCache(long ttl) {
+    this.ttl = ttl;
+  }
+  
+  /**
+   * Cache an entry.
+   * @param method HTTP method
+   * @param path path pattern 
+   * @param userId X-Okapi-User-Id header to cache
+   * @param xokapiPerms X-Okapi-Permissions header to cache
+   * @param token access token to cache
+   */
+  public void put(String tenant, String method, String path, String userId, String xokapiPerms,
+      String keyToken, String token) {
+    long now = System.currentTimeMillis();
+    CacheEntry entry = new CacheEntry(token, userId, xokapiPerms, now + ttl);
+    String key = genKey(method, path, keyToken, userId);
+    MetricsHelper.recordTokenCacheCached(tenant, method, path, userId);
+    logger.info("Caching: {} {}", key, token);
+    cache.put(key, entry);
+  }
```

Comment:
The cache can grow unlimited. Old entries that are expired and not used stay there forever, they are not evicted.
My suggestion:
Convert into a least recently used cache (LRUCache), examples: https://stackoverflow.com/questions/221525/how-would-you-implement-an-lru-cache-in-java
Whenever an entry is put into the cache evict the two oldest entries they are expired. This avoids a timer, is fast, and shrinks the cache if needed.

Reply:
I intentionally kept this super simple for the PoC but fully intend to address these things as part of https://issues.folio.org/browse/OKAPI-890